cpp/src/arrow/python/datetime.h

```diff
@@ -0,0 +1,66 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PYARROW_UTIL_DATETIME_H
+#define PYARROW_UTIL_DATETIME_H
+
+#include <algorithm>
+#include <chrono>
+
+#include "arrow/python/platform.h"
+#include "arrow/status.h"
+#include "arrow/type.h"
+#include "arrow/util/logging.h"
+
+namespace arrow {
+namespace py {
+
+int64_t PyTime_to_us(PyObject* pytime);
```

Comment:
@pitrou apply `ARROW_PYTHON_EXPORT` for all?

Reply:
Well, the functions which were inlined should be kept inline IMO.

Reply:
Then I need to move all of the implementation to the header, including the static functions (practically the whole implementation).

Reply:
That's how it was already, right?

Reply:
If we keep everything as internal than that could work, although I prefer readable header files. 
Actually we should benchmark whether it has a performance impact or not.

Could you recommend me specific benchmarks if we have any?